File: go/dsa/list/SinglyLinkedList/SinglyLinkedList.go
package singlylinkedlist

import (
	"errors"
	"fmt"
)

// ListNode representa um nó na lista encadeada simples.
type ListNode struct {
	Data int        // Valor armazenado no nó.
	Next *ListNode  // Ponteiro para o próximo nó na lista.
}

// SinglyLinkedList representa a lista encadeada simples.
type SinglyLinkedList struct {
	Head *ListNode  // Ponteiro para o primeiro nó da lista.
}

// Display exibe todos os nós da lista encadeada.
func (sll *SinglyLinkedList) Display() {
	current := sll.Head
	for current != nil {
		fmt.Printf("%d --> ", current.Data)
		current = current.Next
	}
	fmt.Println("null")  // Final da lista
}

// Length retorna o comprimento (número de nós) da lista.
func (sll *SinglyLinkedList) Length() int {
	if sll.Head == nil {
		return 0  // Lista vazia
	}
	count := 0
	current := sll.Head
	for current != nil {
		count++
		current = current.Next
	}
	return count  // Retorna o número de nós na lista
}

// InsertFirst insere um nó no início da lista.
func (sll *SinglyLinkedList) InsertFirst(value int) {
	newNode := &ListNode{Data: value}
	newNode.Next = sll.Head  // Aponta para o antigo primeiro nó
	sll.Head = newNode       // Atualiza o Head para o novo nó
}

// Insert insere um nó em uma posição específica na lista (1-indexada).
func (sll *SinglyLinkedList) Insert(position, value int) {
	newNode := &ListNode{Data: value}
	if position == 1 {
		// Inserção no início
		newNode.Next = sll.Head
		sll.Head = newNode
	} else {
		// Inserção em outra posição
		previous := sll.Head
		count := 1
		for count < position-1 {
			previous = previous.Next
			count++
		}
		newNode.Next = previous.Next
		previous.Next = newNode
	}
}

// InsertLast insere um nó no final da lista.
func (sll *SinglyLinkedList) InsertLast(value int) {
	newNode := &ListNode{Data: value}
	if sll.Head == nil {
		// Caso especial: lista vazia
		sll.Head = newNode
		return
	}
	current := sll.Head
	for current.Next != nil {
		current = current.Next
	}
	current.Next = newNode  // Adiciona o novo nó ao final
}

// DeleteFirst remove e retorna o primeiro nó da lista.
func (sll *SinglyLinkedList) DeleteFirst() *ListNode {
	if sll.Head == nil {
		return nil  // Lista vazia, nada para remover
	}
	temp := sll.Head
	sll.Head = sll.Head.Next  // Atualiza o Head para o próximo nó
	temp.Next = nil           // Desconecta o nó removido
	return temp
}

// Delete remove um nó de uma posição específica na lista.
func (sll *SinglyLinkedList) Delete(position int) {
	if position == 1 {
		// Remoção no início
		sll.Head = sll.Head.Next
	} else {
		// Remoção em outra posição
		previous := sll.Head
		count := 1
		for count < position-1 {
			previous = previous.Next
			count++
		}
		current := previous.Next
		previous.Next = current.Next  // Remove o nó
	}
}

// DeleteLast remove e retorna o último nó da lista.
func (sll *SinglyLinkedList) DeleteLast() *ListNode {
	if sll.Head == nil {
		return sll.Head  // Lista vazia
	}
	if sll.Head.Next == nil {
		// Caso especial: lista com um único nó
		temp := sll.Head
		sll.Head = nil
		return temp
	}
	current := sll.Head
	var previous *ListNode
	for current.Next != nil {
		previous = current
		current = current.Next
	}
	previous.Next = nil  // Remove o último nó
	return current
}

// Find verifica se um valor específico existe na lista.
func (sll *SinglyLinkedList) Find(searchKey int) bool {
	current := sll.Head
	for current != nil {
		if current.Data == searchKey {
			return true  // Valor encontrado
		}
		current = current.Next
	}
	return false  // Valor não encontrado
}

// Reverse reverte a ordem dos nós na lista.
func (sll *SinglyLinkedList) Reverse() *ListNode {
	if sll.Head == nil {
		return nil  // Lista vazia
	}

	var previous, Next *ListNode
	current := sll.Head
	for current != nil {
		Next = current.Next    // Armazena o próximo nó
		current.Next = previous  // Inverte a direção do ponteiro
		previous = current     // Move previous e current para frente
		current = Next
	}
	sll.Head = previous  // Atualiza o Head para o novo primeiro nó
	return sll.Head
}

// GetMiddleNode retorna o nó do meio da lista (para listas ímpares) ou o segundo nó do meio (para listas pares).
func (sll *SinglyLinkedList) GetMiddleNode() *ListNode {
	if sll.Head == nil {
		return nil  // Lista vazia
	}
	slowPtr := sll.Head
	fastPtr := sll.Head
	// O ponteiro lento avança uma posição, enquanto o rápido avança duas posições.
	// Quando o ponteiro rápido chega ao final, o ponteiro lento está no meio.
	for fastPtr != nil && fastPtr.Next != nil {
		slowPtr = slowPtr.Next
		fastPtr = fastPtr.Next.Next
	}
	return slowPtr
}

// GetNthNodeFromEnd retorna o n-ésimo nó a partir do final da lista.
func (sll *SinglyLinkedList) GetNthNodeFromEnd(n int) (*ListNode, error) {
	if sll.Head == nil || n <= 0 {
		return nil, errors.New("valor inválido para n")  // Verificação de entrada inválida
	}
	mainPtr, refPtr := sll.Head, sll.Head
	count := 0
	// Mova o ponteiro de referência n posições à frente
	for count < n {
		if refPtr == nil {
			return nil, errors.New("valor inválido para n")
		}
		refPtr = refPtr.Next
		count++
	}
	// Mova ambos os ponteiros até que o ponteiro de referência chegue ao final
	for refPtr != nil {
		refPtr = refPtr.Next
		mainPtr = mainPtr.Next
	}
	return mainPtr, nil
}

// InsertInSortedList insere um nó em uma lista ordenada, mantendo a ordem.
func (sll *SinglyLinkedList) InsertInSortedList(value int) *ListNode {
	newNode := &ListNode{Data: value}
	if sll.Head == nil || sll.Head.Data >= newNode.Data {
		// Caso especial: inserção no início
		newNode.Next = sll.Head
		sll.Head = newNode
		return sll.Head
	}
	current := sll.Head
	// Encontre a posição correta para inserção
	for current.Next != nil && current.Next.Data < newNode.Data {
		current = current.Next
	}
	newNode.Next = current.Next
	current.Next = newNode
	return sll.Head
}

// DeleteNode remove o nó que contém um valor específico.
func (sll *SinglyLinkedList) DeleteNode(key int) {
	current := sll.Head
	var previous *ListNode
	// Caso especial: o nó a ser removido é o primeiro
	if current != nil && current.Data == key {
		sll.Head = current.Next
		return
	}
	// Encontre o nó a ser removido
	for current != nil && current.Data != key {
		previous = current
		current = current.Next
	}
	// Se o nó não for encontrado, faça nada
	if current == nil {
		return
	}
	// Remova o nó da lista
	previous.Next = current.Next
}

// ContainsLoop verifica se a lista contém um loop.
func (sll *SinglyLinkedList) ContainsLoop() bool {
	slowPtr, fastPtr := sll.Head, sll.Head
	// Use dois ponteiros, um movendo-se mais rápido que o outro.
	// Se houver um loop, eles eventualmente se encontrarão.
	for fastPtr != nil && fastPtr.Next != nil {
		slowPtr = slowPtr.Next
		fastPtr = fastPtr.Next.Next
		if slowPtr == fastPtr {
			return true  // Loop detectado
		}
	}
	return false  // Nenhum loop detectado
}

// StartNodeInALoop encontra o nó onde o loop começa, se houver.
func (sll *SinglyLinkedList) StartNodeInALoop() *ListNode {
	slowPtr, fastPtr := sll.Head, sll.Head
	// Primeiro, detecte se há um loop usando dois ponteiros.
	for fastPtr != nil && fastPtr.Next != nil {
		slowPtr = slowPtr.Next
		fastPtr = fastPtr.Next.Next
		if slowPtr == fastPtr {
			// Encontre o nó de início do loop
			return sll.getStartingNode(slowPtr)
		}
	}
	return nil  // Nenhum loop encontrado
}

// getStartingNode encontra o nó de início do loop.
func (sll *SinglyLinkedList) getStartingNode(slowPtr *ListNode) *ListNode {
	temp := sll.Head
	// Move ambos os ponteiros até que eles se encontrem no nó de início do loop.
	for temp != slowPtr {
		temp = temp.Next
		slowPtr = slowPtr.Next
	}
	return temp
}

// RemoveLoop remove um loop na lista, se existir.
func (sll *SinglyLinkedList) RemoveLoop() {
	slowPtr, fastPtr := sll.Head, sll.Head
	// Primeiro, detecte o loop.
	for fastPtr != nil && fastPtr.Next != nil {
		slowPtr = slowPtr.Next
		fastPtr = fastPtr.Next.Next
		if slowPtr == fastPtr {
			// Remova o loop uma vez detectado
			sll.removeLoop(slowPtr)
			return
		}
	}
}

// removeLoop desconecta o nó que forma o loop.
func (sll *SinglyLinkedList) removeLoop(slowPtr *ListNode) {
	temp := sll.Head
	// Encontre o nó onde o loop começa e desconecte-o.
	for temp.Next != slowPtr.Next {
		temp = temp.Next
		slowPtr = slowPtr.Next
	}
	slowPtr.Next = nil  // Remove o loop
}

// CreateALoopInLinkedList cria um loop na lista para fins de teste.
func (sll *SinglyLinkedList) CreateALoopInLinkedList() {
	first := &ListNode{Data: 1}
	second := &ListNode{Data: 2}
	third := &ListNode{Data: 3}
	fourth := &ListNode{Data: 4}
	fifth := &ListNode{Data: 5}
	sixth := &ListNode{Data: 6}

	sll.Head = first
	first.Next = second
	second.Next = third
	third.Next = fourth
	fourth.Next = fifth
	fifth.Next = sixth
	sixth.Next = third // Cria um loop apontando para o terceiro nó
}

// Merge mescla duas listas encadeadas ordenadas em uma única lista ordenada.
func Merge(a, b *ListNode) *ListNode {
	dummy := &ListNode{}
	tail := dummy
	// Continue mesclando até que uma das listas acabe.
	for a != nil && b != nil {
		if a.Data <= b.Data {
			tail.Next = a
			a = a.Next
		} else {
			tail.Next = b
			b = b.Next
		}
		tail = tail.Next
	}
	// Anexa o restante da lista que ainda tem elementos.
	if a != nil {
		tail.Next = a
	} else {
		tail.Next = b
	}
	return dummy.Next  // Retorna a lista mesclada
}

File: csharp/src/dsa/list/SinglyLinkedList.cs
namespace dsa.list;

public class SinglyLinkedList
{
    // static void Main()
    // {
    //     SinglyLinkedList sll1 = new();
    //     sll1.insertLast(1);
    //     sll1.insertLast(4);
    //     sll1.insertLast(8);

    //     SinglyLinkedList sll2 = new();

    //     sll2.insertLast(3);
    //     sll2.insertLast(5);
    //     sll2.insertLast(8);
    //     sll2.insertLast(9);
    //     sll2.insertLast(14);
    //     sll2.insertLast(18);

    //     sll2.display();
    //     sll2.display();

    //     SinglyLinkedList result = new();
    //     result.head = merge(sll1.head, sll2.head);

    //     result.display();
    // }

    // Variável head do tipo ListNode
    public ListNode head;

    public class ListNode
    {
        // Variável para armazenar o valor do nó
        public int data;

        // Variável para referenciar o próximo nó na lista
        public ListNode next;

        public ListNode(int data)
        {
            // Atribui o valor passado para a variável data
            this.data = data;

            // Inicializa o próximo nó como null
            this.next = null;
        }
    }

    public void display()
    {
        // Inicializa o nó atual como o cabeçalho da lista
        ListNode current = head;

        // Loop para percorrer a lista
        while (current != null)
        {
            // Imprime o valor do nó atual
            Console.Write(current.data + "--> ");

            // Move para o próximo nó
            current = current.next;
        }

        // Imprime null no final da lista
        Console.Write("null");


        Console.WriteLine();
    }

    // Método para obter o comprimento da lista
    public int length()
    {
        // Se a lista estiver vazia, retorna 0
        if (head == null)
        {
            return 0;
        }

        // Inicializa o contador de nós
        int count = 0;

        // Inicializa o nó atual como o cabeçalho da lista
        ListNode current = head;

        // Loop para percorrer a lista
        while (current != null)
        {
            // Incrementa o contador
            count++;

            // Move para o próximo nó
            current = current.next;
        }

        // Retorna o número de nós na lista
        return count;
    }


    // Método para inserir um nó no início da lista
    public void insertFirst(int value)
    {
        // Cria um novo nó com o valor fornecido
        ListNode newNode = new ListNode(value);

        // O próximo do novo nó aponta para o atual cabeçalho da lista
        newNode.next = head;

        // O novo nó se torna o cabeçalho da lista
        head = newNode;
    }

    // Método para inserir um nó em uma posição específica na lista
    public void insert(int position, int value)
    {
        // Cria um novo nó com o valor fornecido
        ListNode node = new ListNode(value);

        // Se a posição é 1, insere no início da lista
        if (position == 1)
        {
            node.next = head;
            head = node;
        }
        else
        {
            // Inicializa o nó anterior como o cabeçalho da lista
            ListNode previous = head;
            int count = 1;

            // Loop para encontrar o nó na posição anterior à posição desejada
            while (count < position - 1)
            {
                previous = previous.next;
                count++;
            }

            // O nó atual é o próximo do nó anterior
            ListNode current = previous.next;

            // O próximo do nó anterior aponta para o novo nó
            previous.next = node;

            // O próximo do novo nó aponta para o nó atual
            node.next = current;
        }
    }

    // Método para inserir um nó no final da lista
    public void insertLast(int value)
    {
        // Cria um novo nó com o valor fornecido
        ListNode newNode = new ListNode(value);

        // Se a lista está vazia, o novo nó se torna o cabeçalho da lista
        if (head == null)
        {
            head = newNode;
            return;
        }

        // Inicializa o nó atual como o cabeçalho da lista
        ListNode current = head;

        // Loop para encontrar o último nó da lista
        while (null != current.next)
        {
            current = current.next;
        }

        // O próximo do último nó aponta para o novo nó
        current.next = newNode;
    }

    // Método para remover o primeiro nó da lista
    public ListNode deleteFirst()
    {
        // Se a lista está vazia, retorna null
        if (head == null)
            return null;

        // O nó temporário aponta para o cabeçalho da lista
        ListNode temp = head;

        // O cabeçalho da lista se torna o próximo do nó temporário
        head = head.next;

        // O próximo do nó temporário se torna null
        temp.next = null;

        // Retorna o nó removido
        return temp;
    }


    // Método para remover um nó em uma posição específica na lista
    public void delete(int position)
    {
        // Se a posição é 1, remove o primeiro nó da lista
        if (position == 1)
            head = head.next;
        else
        {
            // Inicializa o nó anterior como o cabeçalho da lista
            ListNode previous = head;
            int count = 1;

            // Loop para encontrar o nó na posição anterior à posição desejada
            while (count < position - 1)
            {
                previous = previous.next;
                count++;
            }

            // O nó atual é o próximo do nó anterior
            ListNode current = previous.next;

            // O próximo do nó anterior aponta para o próximo do nó atual
            previous.next = current.next;
        }
    }

    // Método para remover o último nó da lista
    public ListNode deleteLast()
    {
        // Se a lista está vazia, retorna o cabeçalho da lista
        if (head == null)
            return head;

        // Se a lista tem apenas um nó, remove o nó e retorna
        if (head.next == null)
        {
            ListNode temp = head;
            head = head.next;
            return temp;
        }

        // Inicializa o nó atual como o cabeçalho da lista
        ListNode current = head;
        ListNode previous = null;

        // Loop para encontrar o último nó da lista
        while (current.next != null)
        {
            previous = current;
            current = current.next;
        }

        // O próximo do nó anterior se torna null
        previous.next = null;

        // Retorna o último nó
        return current;
    }

    // Método para encontrar um valor na lista
    public bool find(int searchKey)
    {
        // Se a lista está vazia, retorna false
        if (head == null)
            return false;

        // Inicializa o nó atual como o cabeçalho da lista
        ListNode current = head;

        // Loop para percorrer a lista
        while (current != null)
        {
            // Se o valor do nó atual é igual à chave de pesquisa, retorna true
            if (current.data == searchKey)
                return true;

            // Move para o próximo nó
            current = current.next;
        }

        // Se a chave de pesquisa não foi encontrada, retorna false
        return false;
    }

    // Método para reverter a lista
    public ListNode reverse()
    {
        // Se a lista está vazia, retorna null
        if (head == null)
            return null;

        // Inicializa o nó atual como o cabeçalho da lista
        ListNode current = head;
        ListNode previous = null;
        ListNode next = null;

        // Loop para percorrer a lista
        while (current != null)
        {
            // O próximo nó é o próximo do nó atual
            next = current.next;

            // O próximo do nó atual se torna o nó anterior
            current.next = previous;

            // O nó anterior se torna o nó atual
            previous = current;

            // O nó atual se torna o próximo nó
            current = next;
        }

        // Retorna o nó anterior que é agora o cabeçalho da lista revertida
        return previous;
    }


    // Método para obter o nó do meio da lista
    public ListNode getMiddleNode()
    {
        // Se a lista está vazia, retorna null
        if (head == null)
            return null;

        // Inicializa dois ponteiros, um se move duas vezes mais rápido que o outro
        ListNode slowPtr = head;
        ListNode fastPtr = head;

        // Enquanto o ponteiro rápido não chegar ao final da lista
        while (fastPtr != null && fastPtr.next != null)
        {
            // Move o ponteiro lento um passo
            slowPtr = slowPtr.next;

            // Move o ponteiro rápido dois passos
            fastPtr = fastPtr.next.next;
        }

        // Quando o ponteiro rápido chega ao final da lista, o ponteiro lento está no meio
        return slowPtr;
    }

    // Método para obter o n-ésimo nó a partir do final da lista
    public ListNode getNthNodeFromEnd(int n)
    {
        // Se a lista está vazia, retorna null
        if (head == null)
            return null;

        // Se n é menor ou igual a 0, lança uma exceção
        if (n <= 0)
            throw new InvalidDataException($"Invalid value: n = {n}");

        // Inicializa dois ponteiros
        ListNode mainPtr = head;
        ListNode refPtr = head;

        int count = 0;

        // Move o ponteiro de referência n passos à frente
        while (count < n)
        {
            // Se o ponteiro de referência chegar ao final da lista antes de mover n passos, lança uma exceção
            if (refPtr == null)
                throw new InvalidDataException($"Invalid value: n = {n}");

            refPtr = refPtr.next;
            count++;
        }

        // Move ambos os ponteiros até que o ponteiro de referência chegue ao final da lista
        while (refPtr != null)
        {
            refPtr = refPtr.next;
            mainPtr = mainPtr.next;
        }

        // Quando o ponteiro de referência chega ao final da lista, o ponteiro principal está no n-ésimo nó a partir do final
        return mainPtr;
    }

    // Método para inserir um nó em uma lista ordenada
    public ListNode insertInSortedList(int value)
    {
        // Cria um novo nó com o valor fornecido
        ListNode newNode = new ListNode(value);

        // Se a lista está vazia, retorna o novo nó
        if (head == null)
            return newNode;

        // Inicializa dois ponteiros
        ListNode current = head;
        ListNode temp = null;

        // Encontra a posição para inserir o novo nó
        while (current != null && current.data < newNode.data)
        {
            temp = current;
            current = current.next;
        }

        // Insere o novo nó na posição correta
        newNode.next = current;
        temp.next = newNode;

        // Retorna o cabeçalho da lista
        return head;
    }

    // Método para remover um nó com uma chave específica
    public void deleteNode(int key)
    {
        // Inicializa dois ponteiros
        ListNode current = head;
        ListNode temp = null;

        // Se a chave está no primeiro nó, remove o primeiro nó
        if (current != null && current.data == key)
        {
            head = current.next;
            return;
        }

        // Encontra o nó com a chave
        while (current != null && current.data != key)
        {
            temp = current;
            current = current.next;
        }

        // Se a chave não foi encontrada, retorna
        if (current == null)
            return;

        // Remove o nó com a chave
        temp.next = current.next;
    }

    // Método para verificar se a lista contém um loop
    public bool containsLoop()
    {
        // Inicializa dois ponteiros, um se move duas vezes mais rápido que o outro
        ListNode fastPtr = head;
        ListNode slowPtr = head;

        // Enquanto o ponteiro rápido não chegar ao final da lista
        while (fastPtr != null && fastPtr.next != null)
        {
            // Move o ponteiro lento um passo
            slowPtr = slowPtr.next;

            // Move o ponteiro rápido dois passos
            fastPtr = fastPtr.next.next;

            // Se os dois ponteiros se encontram, a lista contém um loop
            if (fastPtr == slowPtr)
                return true;
        }

        // Se o ponteiro rápido chega ao final da lista, a lista não contém um loop
        return false;
    }

    // Método para encontrar o nó de início de um loop em uma lista
    public ListNode startNodeInALoop()
    {
        // Inicializa dois ponteiros, um se move duas vezes mais rápido que o outro
        ListNode fastPtr = head;
        ListNode slowPtr = head;

        // Enquanto o ponteiro rápido não chegar ao final da lista
        while (fastPtr != null && fastPtr.next != null)
        {
            // Move o ponteiro lento um passo
            slowPtr = slowPtr.next;

            // Move o ponteiro rápido dois passos
            fastPtr = fastPtr.next.next;

            // Se os dois ponteiros se encontram, a lista contém um loop
            if (fastPtr == slowPtr)
                // Retorna o nó de início do loop
                return getStartingNode(slowPtr);
        }

        // Se o ponteiro rápido chega ao final da lista, a lista não contém um loop
        return null;
    }

    // Método para encontrar o nó de início de um loop
    public ListNode getStartingNode(ListNode slowPtr)
    {
        // Inicializa um ponteiro temporário no cabeçalho da lista
        ListNode temp = head;

        // Enquanto o ponteiro temporário e o ponteiro lento não se encontram
        while (temp != slowPtr)
        {
            // Move ambos os ponteiros um passo
            temp = temp.next;
            slowPtr = slowPtr.next;
        }

        // Retorna o nó de início do loop
        return temp;
    }

    // Método para remover um loop de uma lista
    public void removeLoop()
    {
        // Inicializa dois ponteiros, um se move duas vezes mais rápido que o outro
        ListNode fastPtr = head;
        ListNode slowPtr = head;

        // Enquanto o ponteiro rápido não chegar ao final da lista
        while (fastPtr != null && fastPtr.next != null)
        {
            // Move o ponteiro lento um passo
            slowPtr = slowPtr.next;

            // Move o ponteiro rápido dois passos
            fastPtr = fastPtr.next.next;

            // Se os dois ponteiros se encontram, a lista contém um loop
            if (fastPtr == slowPtr)
            {
                // Remove o loop
                removeLoop(slowPtr);
                return;
            }
        }
    }

    // Método para remover um loop de uma lista
    public void removeLoop(ListNode slowPtr)
    {
        // Inicializa um ponteiro temporário no cabeçalho da lista
        ListNode temp = head;

        // Enquanto o próximo do ponteiro temporário e o próximo do ponteiro lento não são iguais
        while (temp.next != slowPtr.next)
        {
            // Move ambos os ponteiros um passo
            temp = temp.next;
            slowPtr = slowPtr.next;
        }

        // Faz o próximo do ponteiro lento null para remover o loop
        slowPtr.next = null;
    }


    // Método para criar um loop em uma lista vinculada
    public void createALoopInLinkedList()
    {
        // Cria seis novos nós
        ListNode first = new ListNode(1);
        ListNode second = new ListNode(2);
        ListNode third = new ListNode(3);
        ListNode fourth = new ListNode(4);
        ListNode fifth = new ListNode(5);
        ListNode sixth = new ListNode(6);

        // Conecta os nós para formar uma lista vinculada
        head = first;
        first.next = second;
        second.next = third;
        third.next = fourth;
        fourth.next = fifth;
        fifth.next = sixth;

        // Cria um loop na lista vinculada fazendo o último nó apontar para o terceiro nó
        sixth.next = third;
    }

    // Método para mesclar duas listas vinculadas ordenadas
    public static ListNode merge(ListNode a, ListNode b)
    {
        // Cria um nó dummy para servir como cabeçalho temporário da lista mesclada
        ListNode dummy = new(0);
        ListNode tail = dummy;

        // Loop para percorrer ambas as listas até que uma delas esteja vazia
        while (a != null && b != null)
        {
            // Se o valor do nó a é menor ou igual ao valor do nó b
            if (a.data <= b.data)
            {
                // Adiciona o nó a à lista mesclada
                tail.next = a;

                // Move para o próximo nó na lista a
                a = a.next;
            }
            else
            {
                // Adiciona o nó b à lista mesclada
                tail.next = b;

                // Move para o próximo nó na lista b
                b = b.next;
            }

            // Move para o próximo nó na lista mesclada
            tail = tail.next;
        }

        // Se a lista a está vazia, adiciona o restante da lista b à lista mesclada
        if (a == null)
            tail.next = b;

        // Se a lista b está vazia, adiciona o restante da lista a à lista mesclada
        else
            tail.next = a;

        // Retorna a lista mesclada, ignorando o nó dummy no início
        return dummy.next;
    }

}
